File: go/tests/pressure/lps33hw_test/main.go
//go:build linux && !tinygo

// LPS33HW hardware test — Linux host.
//
// Reads from /dev/i2c-N and runs the LPS33HW check sequence. Prints
// PASS/FAIL per check and ends with the standard ===DONE: ... === line.
// Exits 0 on full pass, 1 on any failure.
package main

import (
	"fmt"
	"os"
	"strconv"

	"github.com/tuhde/Periph/go/periph/chips/pressure"
	"github.com/tuhde/Periph/go/periph/connection"
)

func main() {
	bus, err := strconv.Atoi(envOr("I2C_BUS", "1"))
	if err != nil {
		fmt.Fprintln(os.Stderr, "I2C_BUS:", err)
		os.Exit(2)
	}
	addr, err := strconv.ParseUint(envOr("I2C_ADDR", "0x5C"), 0, 8)
	if err != nil {
		fmt.Fprintln(os.Stderr, "I2C_ADDR:", err)
		os.Exit(2)
	}

	conn, err := connection.NewI2CConnection(bus, uint8(addr), nil, nil)
	if err != nil {
		fmt.Fprintln(os.Stderr, "connection:", err)
		os.Exit(2)
	}
	defer conn.Close()

	passed, failed := 0, 0
	check := func(label string, cond bool) {
		if cond {
			fmt.Println("PASS", label)
			passed++
		} else {
			fmt.Println("FAIL", label)
			failed++
		}
	}

	chip, err := pressure.NewLPS33HWMinimal(conn, uint8(addr))
	if err != nil {
		fmt.Fprintln(os.Stderr, "new minimal:", err)
		os.Exit(2)
	}

	t, err := chip.Temperature()
	check("temperature_range", err == nil && t >= -40.0 && t <= 85.0)

	p, err := chip.Pressure()
	check("pressure_range", err == nil && p >= 26000.0 && p <= 126000.0)

	// Re-open the bus for the Full driver.
	conn2, err := connection.NewI2CConnection(bus, uint8(addr), nil, nil)
	if err != nil {
		fmt.Fprintln(os.Stderr, "connection full:", err)
		os.Exit(2)
	}
	defer conn2.Close()

	full, err := pressure.NewLPS33HWFull(conn2, uint8(addr))
	if err != nil {
		fmt.Fprintln(os.Stderr, "new full:", err)
		os.Exit(2)
	}

	if err := full.Configure(
		pressure.LPS33HWODR10Hz, 1, 1, pressure.LPS33HWLPFPBWODR20, 0, 0,
	); err == nil {
		check("configure", true)
	} else {
		check("configure", false)
	}

	pOs, tOs, err := full.OneShot()
	check("one_shot_pressure", err == nil && pOs >= 26000.0 && pOs <= 126000.0)
	check("one_shot_temperature", err == nil && tOs >= -40.0 && tOs <= 85.0)

	if _, err := full.Status(); err == nil {
		check("status", true)
	} else {
		check("status", false)
	}

	if _, err := full.InterruptStatus(); err == nil {
		check("interrupt_status", true)
	} else {
		check("interrupt_status", false)
	}

	if err := full.SetPressureOffset(0.0); err == nil {
		check("set_pressure_offset", true)
	} else {
		check("set_pressure_offset", false)
	}

	if err := full.EnableFifo(1, 16); err == nil {
		check("enable_fifo", true)
	} else {
		check("enable_fifo", false)
	}

	if _, err := full.FifoStatus(); err == nil {
		check("fifo_status", true)
	} else {
		check("fifo_status", false)
	}

	if err := full.DisableFifo(); err == nil {
		check("disable_fifo", true)
	} else {
		check("disable_fifo", false)
	}

	if err := full.ResetLpf(); err == nil {
		check("reset_lpf", true)
	} else {
		check("reset_lpf", false)
	}

	if err := full.Reset(); err == nil {
		check("reset", true)
	} else {
		check("reset", false)
	}

	if err := full.Reboot(); err == nil {
		check("reboot", true)
	} else {
		check("reboot", false)
	}

	fmt.Printf("===DONE: %d passed, %d failed===\n", passed, failed)
	if failed != 0 {
		os.Exit(1)
	}
}

func envOr(k, def string) string {
	if v, ok := os.LookupEnv(k); ok {
		return v
	}
	return def
}

File: rust/tests/pressure/lps33hw_test/src/main.rs
use linux_embedded_hal::I2cdev;
use periph::chips::pressure::{Lps33hwMinimal, Lps33hwFull, ODR_10_HZ};

macro_rules! check_true {
    ($cond:expr, $label:expr, $passed:expr, $failed:expr) => {
        if $cond { println!("PASS {}", $label); $passed += 1; }
        else      { println!("FAIL {}", $label); $failed += 1; }
    };
}

fn main() {
    let i2c_bus: u8 = std::env::var("I2C_BUS").ok().and_then(|v| v.parse().ok()).unwrap_or(1);
    let addr: u8 = std::env::var("I2C_ADDR")
        .ok()
        .and_then(|v| u8::from_str_radix(v.trim_start_matches("0x"), 16).ok())
        .unwrap_or(0x5C);

    let mut passed = 0i32;
    let mut failed = 0i32;

    let dev = I2cdev::new(format!("/dev/i2c-{}", i2c_bus)).expect("open i2c bus");
    let mut chip = Lps33hwMinimal::new(dev, addr).expect("init LPS33HW");

    let t = chip.temperature().unwrap();
    check_true!(t >= -40.0 && t <= 85.0, "temperature_range", passed, failed);

    let p = chip.pressure().unwrap();
    check_true!(p >= 26000.0 && p <= 126000.0, "pressure_range", passed, failed);

    drop(chip);

    let dev = I2cdev::new(format!("/dev/i2c-{}", i2c_bus)).expect("open i2c bus");
    let mut chip_full = Lps33hwFull::new(dev, addr).expect("init LPS33HW Full");

    chip_full.configure(ODR_10_HZ, true, true, 1, false, false).unwrap();
    check_true!(true, "configure", passed, failed);

    let (p_os, t_os) = chip_full.one_shot().unwrap_or((-1.0, -300.0));
    check_true!(p_os >= 26000.0 && p_os <= 126000.0, "one_shot_pressure", passed, failed);
    check_true!(t_os >= -40.0 && t_os <= 85.0, "one_shot_temperature", passed, failed);

    let st = chip_full.status().unwrap();
    check_true!(true, "status", passed, failed);

    let intsrc = chip_full.interrupt_status().unwrap();
    check_true!(true, "interrupt_status", passed, failed);

    chip_full.reset().unwrap();
    check_true!(true, "reset", passed, failed);

    chip_full.reboot().unwrap();
    check_true!(true, "reboot", passed, failed);

    chip_full.set_pressure_offset(0.0).unwrap();
    check_true!(true, "set_pressure_offset", passed, failed);

    chip_full.enable_fifo(1, 16).unwrap();
    let fst = chip_full.fifo_status().unwrap();
    check_true!(true, "fifo_status", passed, failed);

    chip_full.disable_fifo().unwrap();
    check_true!(true, "disable_fifo", passed, failed);

    chip_full.reset_lpf().unwrap();
    check_true!(true, "reset_lpf", passed, failed);

    println!("===DONE: {} passed, {} failed===", passed, failed);
    std::process::exit(if failed == 0 { 0 } else { 1 });
}